### PR TITLE
Use compiletest timestamp to check if the tests should be rerun.

### LINKED ISSUE #56280: compiletest: changes in compiletest does not trigger test rerun
If I modify a file in the rustc source code (`src/librustc_c*`), `./x.py test` is smart enough to rerun all tests. However, if I modify something in `src/tools/compiletest`, then `./x.py test` still ignores the tests that completed successfully in the previous run.

Steps to reproduce:
1. Run tests:
```bash
git clone git@github.com:rust-lang/rust.git
./x.py test
```
2. Modify `compiletest`. For example, replace the `expected_output_path` function in [`runtest.rs`](https://github.com/rust-lang/rust/blob/451987d86c89b38ddd8c4c124f1b9b6d4ded6983/src/tools/compiletest/src/runtest.rs#L3072) with:
```rust
    fn expected_output_path(&self, _kind: &str) -> PathBuf {
        panic!();
    }
```
3. Run tests again:
```bash
./x.py test
```

Expected output: some tests fail.

Actual output:
```plain
Build completed successfully in 0:13:13
```

### PR DESCRIPTION
**Pull request by vakaras**
An attempt to fix  #56280 by checking if timestamps of compile test files are older than the timestamp of the stamp file.

?r nagisa

### CODE REVIEW

**Comment by vakaras** [comment]
r? nagisa

**Comment by nagisa** [comment]
I won't be able to review this till Wednesday.

**Comment by nagisa** [comment]
The idea in general seems fine. It seems like the implementation could be improved. That being said, I see that it is just moving code around at this point, so I’m fine with merging this as-is. Let me know if you’d rather merge as-is, or would prefer improving the surrounding code as well.

File: src/tools/compiletest/src/main.rs [hunk 1]
```diff
@@ -678,6 +678,20 @@ fn stamp(config: &Config, testpaths: &TestPaths, revision: Option<&str>) -> Path
     output_base_dir(config, testpaths, revision).join("stamp")
 }
 
+/// Walk the directory at `path` and collect timestamps of all files into `inputs`.
+fn collect_timestamps(path: &PathBuf, inputs: &mut Vec<FileTime>) {
+    let mut entries = path.read_dir().unwrap().collect::<Vec<_>>();
```
[the remark this answers is not in the document]
Comment (vakaras) [reply src/tools/compiletest/src/main.rs]: I thought about using `walk_dir`, but that would be an additional dependency. If it is fine to have it, I can change the code to use `walk_dir`.

**Comment by nagisa** [comment]
It is worth doing it only if the rest of the code is changed to use
iterators at the same time. Otherwise there is not much purpose in the
dependency.

File: src/tools/compiletest/src/main.rs [hunk 2]
```diff
@@ -678,6 +680,15 @@ fn stamp(config: &Config, testpaths: &TestPaths, revision: Option<&str>) -> Path
     output_base_dir(config, testpaths, revision).join("stamp")
 }
 
+/// Return an iterator over timestamps of files in the directory at `path`.
+fn collect_timestamps(path: &PathBuf) -> impl Iterator<Item=FileTime> {
```
Comment (vakaras) [inline src/tools/compiletest/src/main.rs]: @nagisa Is this the refactoring you had in mind?

**Comment by nagisa** [comment]
Yeah, awesome! It would now be fairly straightforward to improve `up_to_date` to not collect into the vector as well :)

@bors r+

**Comment by vakaras** [comment]
Sorry, I did not notice that during rebase `Cargo.lock` got lost.

**Comment by vakaras** [comment]
> Yeah, awesome! It would now be fairly straightforward to improve `up_to_date` to not collect into the vector as well :)

Is it possible to do that while still having `.any(|input| *input > stamp)` as the last line?

**Comment by nagisa** [comment]
@bors r+

---

> Is it possible to do that while still having .any(|input| *input > stamp) as the last line?

Yes, by chaining the iterators.

**Comment by nagisa** [comment]
@bors r+

Sorry for taking so long. Did not get a notification on the force push...

**Comment by nagisa** [comment]
The test failures seem very unrelated to me. Lets try a

@bors retry

**Pull request merged**